feat: add priority queue to reconciling work in apply_controller

sudheerv opened the pull request:
### Description of your changes

[isssue#46]

add priority queue to reconciling work in apply_controller

apply_controllers in the memberagent currently watches all the
Work resources created in the designated namespace on the Hub
cluster and reconciles them in order. This can result in serious
bottleneck in applying newer resources or newer updates to
existing resources over time especially when there are 1000s
of namespaces created a long time ago and never have any
updates to them, yet reconciled one by one before the
controller can get to the newer ones.

Fixes #

This PR integrates a priority queue for adding work objects to
the watch list to prioritize work objects that are newer, have
recent updates or have their conditions not Ready yet.

Note that the PR uses both `For` and `Watches` in the controller
setup. This is needed to avoid a runtime panic when using just `Watches`

This can also be done using just `For` without `Watches`, but that model 
would not use a `PriorityQueue` and just plays directly with the events. 
With that model, we will only watch the changed/new work objects and 
completely ignore old objects. With the `PriorityQueue`, we are actually 
adding every single Work object to the queue just in a prioritized order.

Signed-off-by: Sudheer Vinukonda <sudheerv@apache.org>

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested
tested it in our production environment

Namespace Readiness (without patch)
```
test-sudheer-ns-10: NamespaceCreated: 2025-04-09T21:47:33Z, AllReady: 2025-04-09T22:13:15Z
test-sudheer-ns-11: NamespaceCreated: 2025-04-09T22:24:32Z, AllReady: 2025-04-09T23:03:57Z
test-sudheer-ns-12: NamespaceCreated: 2025-04-09T22:24:47Z, AllReady: 2025-04-09T23:03:57Z
test-sudheer-ns-13: NamespaceCreated: 2025-04-09T22:24:56Z, AllReady: 2025-04-09T23:03:57Z
```
Namespace Readiness (with patch)
```
test-sudheer-ns-37: NamespaceCreated: 2025-04-11T17:30:46Z, AllReady: 2025-04-11T17:31:16Z
test-sudheer-ns-38: NamespaceCreated: 2025-04-11T17:30:57Z, AllReady: 2025-04-11T17:31:27Z
test-sudheer-ns-39: NamespaceCreated: 2025-04-11T17:31:07Z, AllReady: 2025-04-11T17:31:37Z
test-sudheer-ns-40: NamespaceCreated: 2025-04-11T17:31:17Z, AllReady: 2025-04-11T17:31:48Z
```
<!--
Before reviewers can be confident in the correctness of this pull request, it needs to tested and shown to be correct.
Briefly describe the testing that has already been done or which is planned for this change.
-->


### Special notes for your reviewer

<!--

Be sure to direct your reviewers' attention to anything that needs special consideration.

-->

sudheerv commented:
@ryanzhang-oss  Can you take a look at the PR pls? Some checks are failing - is there a format for the title that I should follow?

ryanzhang-oss commented on pkg/controllers/work/apply_controller.go:
```diff
@@ -101,6 +107,102 @@ const (
+	// check if the object has been recently modified
+	availCondLastUpdatedTime := availCond.LastTransitionTime.Time
+	appliedCondLastUpdatedTime := appliedCond.LastTransitionTime.Time
+	if time.Since(availCondLastUpdatedTime) < 1*time.Hour || time.Since(appliedCondLastUpdatedTime) < 1*time.Hour {
```
```suggestion
	if time.Since(availCondLastUpdatedTime) < workAgeToReconcile  || time.Since(appliedCondLastUpdatedTime) < workAgeToReconcile  {
```

sudheerv replied:
updated.

ryanzhang-oss commented on pkg/controllers/work/apply_controller.go:
```diff
@@ -698,7 +812,8 @@ func (r *ApplyWorkReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		}).
-		For(&fleetv1beta1.Work{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		For(&fleetv1beta1.Work{}).
+		Watches(&fleetv1beta1.Work{}, eventHandler).
```
This handler does not seem to filter out status change? That will be a lot more noises.

sudheerv replied:
added filter

ryanzhang-oss reviewed (commented):
Can you add a feature flag for that is turned off by default?

sudheerv commented:
> Can you add a feature flag (that is turned off by default) for this feature?

Added. Turns out there is duplication in the code on the HEAD branch (between work/apply_controller.go and workapplier/controller.go. I missed that in the original PR since v0.9.5 did not have this. I'm presuming there is a planned cleanup for work/apply_controller.go but in this PR, I added the code to both controllers. Plmk if this is not the correct understanding.

sudheerv commented on pkg/controllers/work/apply_controller.go:
```diff
@@ -101,6 +107,102 @@ const (
+		Name:      name,
+	}, &work)
+	if err != nil {
+		// Log and return
```
I was on two minds, since trying to add the object to queue may not really help in this case since the object GET failed but changed it to keep the semantics consistent.

ryanzhang-oss commented on pkg/controllers/workapplier/controller.go:
```diff
@@ -71,17 +77,135 @@ const (
 )
 
+const (
+	workAgeToReconcile = 1 * time.Hour
```
can this also be a parameter with a default?

sudheerv replied:
Ah, makes sense. Updated (only updated the workapplier/controller.go and not the work/apply_controller.go)

ryanzhang-oss commented on pkg/controllers/workapplier/controller.go:
```diff
@@ -483,6 +614,26 @@ func (r *Reconciler) Leave(ctx context.Context) error {
+
+	if r.watchWorkWithPriorityQueue {
+		workAgeToReconcile = time.Duration(r.watchWorkReconcileAgeMinutes) * time.Minute
+		return ctrl.NewControllerManagedBy(mgr).Named("work-controller").
```
```suggestion
		return ctrl.NewControllerManagedBy(mgr).Named("work-applier-controller").
```

ryanzhang-oss commented on pkg/controllers/workapplier/controller.go:
```diff
@@ -483,6 +614,26 @@ func (r *Reconciler) Leave(ctx context.Context) error {
+			WithOptions(ctrloption.Options{
+				MaxConcurrentReconciles: r.concurrentReconciles,
+			}).
+			For(&fleetv1beta1.Work{}).
+			Watches(&fleetv1beta1.Work{}, eventHandler).
```
are these not duplicate? Do you still want to watch for status change?

sudheerv replied:
>> are these not duplicate? 

Without using `For(&fleetv1beta1.Work{}).`, `Watches(&fleetv1beta1.Work{}, eventHandler).` causes a runtime panic. 

Here's what chatgpt says

```
			For(&fleetv1beta1.Work{}).
			Watches(&fleetv1beta1.Work{}, eventHandler).

without using For, there is a runtime panic. Why are both needed

ChatGPT said:
In Kubernetes controller-runtime, using both:

go
Copy
Edit
For(&fleetv1beta1.Work{}).
Watches(&fleetv1beta1.Work{}, eventHandler).
might look redundant, but they serve different purposes, and omitting For(...) can lead to a runtime panic — exactly as you observed.

⚙️ Why For(...) Is Required
go
Copy
Edit
For(&fleetv1beta1.Work{})
This tells the controller what the primary resource is — the kind it Reconciles. It's required for the controller to set up the internal machinery that connects:

the GVK (Group-Version-Kind) of Work

to your Reconciler

and creates the initial informers and caches behind the scenes.

If you remove For(...), controller-runtime doesn’t know what kind of object your Reconciler is responsible for, so when events come in (even via Watches(...)), it doesn't know how to hydrate them correctly, leading to nil objects or panics during reconciliation.

🧭 What Watches(...) Does
go
Copy
Edit
Watches(&fleetv1beta1.Work{}, eventHandler)
This adds an additional watch on the same or different resource with a custom event handler, often to control:

When objects should be enqueued

Custom logic like filtering or prioritization

Watching secondary resources (e.g., Secrets, Pods related to a Work)

In your case, you're watching the same resource (Work) twice:

Once via For(...) (default watch that enqueues on every change)

Again via Watches(...) with a custom eventHandler (maybe for fine-grained control, e.g., filtering status-only changes or using a priority queue)

This is a valid and common pattern when you need to override default behavior.
```



re: status change, I'm ignoring the status changes via the Update function below.

```
func (h *PriorityQueueEventHandler) Update(ctx context.Context, evt event.TypedUpdateEvent[client.Object], queue workqueue.TypedRateLimitingInterface[reconcile.Request]) {
	// Ignore updates where only the status changed
	oldObj := evt.ObjectOld.DeepCopyObject()
	newObj := evt.ObjectNew.DeepCopyObject()

	// Zero out the status
	if oldWork, ok := oldObj.(*fleetv1beta1.Work); ok {
		oldWork.Status = fleetv1beta1.WorkStatus{}
	}
	if newWork, ok := newObj.(*fleetv1beta1.Work); ok {
		newWork.Status = fleetv1beta1.WorkStatus{}
	}

	if !equality.Semantic.DeepEqual(oldObj, newObj) {
		// ignore status changes to prevent noise
		h.AddToPriorityQueue(ctx, evt.ObjectNew, true)
	}
	klog.V(4).InfoS("ignoring update event with only status change", "work", evt.ObjectNew.GetName())
}

```
